pkg/daemon/controller/controller.go

```diff
@@ -64,58 +77,133 @@ func RunController(kubeconfigPath string) error {
+	ctx := ctrl.SetupSignalHandler()
 	var nodes core.NodeList
-	err = clientset.List(context.TODO(), &nodes)
+	err = directClient.List(ctx, &nodes)
 	if err != nil {
 		return err
 	}
 	addrs, err := localInterfaceAddresses()
 	if err != nil {
 		return err
 	}
-	node, err := currentNode(&nodes, addrs)
+	node, err := findNodeByAddress(&nodes, addrs)
+	if err != nil {
+		return err
+	}
+
+	ctrlMgr, err := ctrl.NewManager(config, ctrl.Options{
+		Namespace: watchNamespace,
+		Scheme:    clientScheme,
+	})
 	if err != nil {
+		return errors.Wrap(err, "unable to start manager")
+	}
+	sc := NewServiceController(ctx, ctrlMgr.GetClient(), svcMgr, node.Name)
+	if err = sc.SetupWithManager(ctrlMgr); err != nil {
 		return err
 	}
+	klog.Info("Starting manager, awaiting events")
+	if err := ctrlMgr.Start(ctx); err != nil {
+		return err
+	}
+	return nil
+}
 
-	desiredVersion, present := node.Annotations["windowsmachineconfig.openshift.io/desired-version"]
+// NewServiceController returns a pointer to a ServiceController object
+func NewServiceController(ctx context.Context, client client.Client, mgr winsvc.Mgr, nodeName string) *ServiceController {
+	return &ServiceController{client: client, Mgr: mgr, ctx: ctx, nodeName: nodeName,
+		watchNamespace: watchNamespace}
+}
+
+// SetupWithManager sets up the controller with the Manager.
+func (sc *ServiceController) SetupWithManager(mgr ctrl.Manager) error {
+	nodePredicate := predicate.Funcs{
+		// A node's name will never change, so it is fine to use the name for node identification
+		// The node must have a desired-version annotation for it to be reconcilable
+		CreateFunc: func(e event.CreateEvent) bool {
+			return sc.nodeName == e.Object.GetName() && e.Object.GetAnnotations()[desiredVersionAnnotation] != ""
+		},
+		UpdateFunc: func(e event.UpdateEvent) bool {
+			// Only process update events if the desired version has changed
+			return sc.nodeName == e.ObjectNew.GetName() &&
+				e.ObjectOld.GetAnnotations()[desiredVersionAnnotation] != e.ObjectNew.GetAnnotations()[desiredVersionAnnotation]
+		},
+		GenericFunc: func(e event.GenericEvent) bool {
+			return sc.nodeName == e.Object.GetName() && e.Object.GetAnnotations()[desiredVersionAnnotation] != ""
+		},
+		DeleteFunc: func(e event.DeleteEvent) bool {
+			return false
+		},
+	}
+	cmPredicate := predicate.NewPredicateFuncs(func(object client.Object) bool {
+		return strings.HasPrefix(object.GetName(), servicescm.NamePrefix)
+	})
+
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&core.Node{}, builder.WithPredicates(nodePredicate)).
+		Watches(&source.Kind{Type: &core.ConfigMap{}}, handler.EnqueueRequestsFromMapFunc(sc.mapToCurrentNode),
+			builder.WithPredicates(cmPredicate)).
+		Complete(sc)
+}
+
+// mapToCurrentNode maps all events to the node associated with this Windows instance
+func (sc *ServiceController) mapToCurrentNode(_ client.Object) []reconcile.Request {
+	return []reconcile.Request{{types.NamespacedName{Name: sc.nodeName}}}
```

Comment:
How do this works when there is more than one ConfigMap with prefix `servicescm.NamePrefix` ?

Reply:
This just makes it so anytime there is a service ConfigMap event that makes it through the predicate, the current VM's node is reconciled. There being multiple ConfigMaps should only happen during upgrades, and won't be a problem.